pkg/integrations/v2/app_o11y_receiver/sourcemaps/sourcemaps.go

```diff
@@ -0,0 +1,331 @@
+	//attempt to find in fs
+	for _, fileconf := range store.config.FileSystem {
+		content, sourceMapURL, err = store.getSourceMapFromFileSystem(sourceURL, release, fileconf)
+		if content != nil || err != nil {
+			return content, sourceMapURL, err
+		}
+	}
+
+	//attempt to download
+	if strings.HasPrefix(sourceURL, "http") && utils.URLMatchesOrigins(sourceURL, store.config.DownloadFromOrigins) {
+		return store.downloadSourceMapContent(sourceURL)
+	}
+	return nil, "", nil
+}
+
+func (store *RealSourceMapStore) getSourceMap(sourceURL string, release string) (*sourceMap, error) {
+	store.Lock()
+	defer store.Unlock()
+
+	cacheKey := fmt.Sprintf("%s__%s", sourceURL, release)
+
+	if smap, ok := store.cache[cacheKey]; ok {
+		return smap, nil
+	}
+	content, sourceMapURL, err := store.getSourceMapContent(sourceURL, release)
+	if err != nil || content == nil {
+		store.cache[cacheKey] = nil
+		return nil, err
+	}
+	if content != nil {
+		consumer, err := sourcemap.Parse(sourceMapURL, content)
+		if err != nil {
+			store.cache[cacheKey] = nil
+			level.Debug(store.l).Log("msg", "failed to parse sourcemap", "url", sourceMapURL, "release", release, "err", err)
+			return nil, err
+		}
+		level.Info(store.l).Log("msg", "successfully parsed sourcemap", "url", sourceMapURL, "release", release)
+		smap := &sourceMap{
+			consumer: consumer,
+		}
+		store.cache[cacheKey] = smap
+		store.metrics.cacheSize.WithLabelValues(getOrigin(sourceURL)).Inc()
+		return smap, nil
+	}
+	return nil, nil
+}
+
+// ResolveSourceLocation resolves minified source location to original source location
+func (store *RealSourceMapStore) ResolveSourceLocation(frame *models.Frame, release string) (*models.Frame, error) {
+	smap, err := store.getSourceMap(frame.Filename, release)
+	if err != nil {
+		return nil, err
+	}
+	if smap == nil {
+		return nil, nil
+	}
+
+	file, function, line, col, ok := smap.consumer.Source(frame.Lineno, frame.Colno)
+	if !ok {
+		return nil, nil
+	}
+	// unfortunately in many cases go-sourcemap fails to determine the original function name.
+	// not a big issue as long as file, line and column are correct
+	if len(function) == 0 {
+		function = "?"
+	}
+	return &models.Frame{
+		Filename: file,
+		Lineno:   line,
+		Colno:    col,
+		Function: function,
+	}, nil
+}
+
+// TransformException will attempt to resolved all monified source locations in the stacktrace with original source locations
+func (store *RealSourceMapStore) TransformException(ex *models.Exception, release string) *models.Exception {
+	if ex.Stacktrace == nil {
+		return ex
+	}
+	frames := []models.Frame{}
+
+	for _, frame := range ex.Stacktrace.Frames {
+		mappedFrame, err := store.ResolveSourceLocation(&frame, release)
+		if err != nil {
+			level.Error(store.l).Log("msg", "Error resolving stack trace frame source location", "err", err)
+			frames = append(frames, frame)
+		} else {
+			if mappedFrame != nil {
```

Comment:
This can be lowered a level of indentation with `else if mappedFrame != nil { } .... else { }`

Reply:
- [ ] refac